docs/manual/src/udl/delegates.md

```diff
@@ -0,0 +1,242 @@
+# Delegate objects
+
+We'll be considering the UDL for an object called `DemoRustObject`:
+
+```webidl
+interface DemoRustObject {
+    void do_expensive_thing();
+    [Throws=RustyError]
+    void do_crashy_thing();
+    string get_state_at(i32 index);
+    void set_state_at(i32 index, string value);
+}
+```
+
+`DemoRustObject` is a `struct` written in Rust, and uniffi generates a class of that name in the foreign languages which forward method calls to the Rust implementation.
+
+## Problem
+
+Just by the names of these methods, we can see that the application might not want to deal with the `DemoRustObject` by itself, without some proper care taken while calling its methods.
+
+A common pattern when using uniffied code is to run some common, but app-specific, code before or after calling the Rust code.
+
+```kotlin
+class DemoLib(
+    val backgroundScope: CoroutineContext,
+    val errorReporter: (e: Exception) -> Unit,
+) {
+    val demoRustObject = new DemoRustObject()
+
+    fun doExpensiveThing() {
+        backgroundScope.launch {
+            demoRustObject.doExpensiveThing()
+        }
+    }
+
+    fun doCrashyThing() {}
+        try {
+            demoRustObject.doCrashyThing()
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+    }
+
+    fun getStateAt(index: Int): String =
+        try {
+            demoRustObject.getState(index)
+        } catch (e: Exception) {
+            errorReporter(e)
+        }
+
+    fun setStateAt(index: Int, value: String) {
+        try {
+            demoRustObject.getState(index)
+        } finally {
+            // not shown here.
+            // Does something with the DemoRustObject
+            this.notifyListeners()
+        }
+    }
+}
+```
+
+This causes a proliferation of boiler plate code. Worse, everytime we add a new method to the `DemoRustObject`, handwritten foreign language code needs to be written to expose it safely.
+
+The more methods that want to do similar things, the more repetitive this gets.
+
+We could isolate the repeated code into a delegate class with `onBackgroundThread` and `withErrorReporter` methods.
+
+We'll handwrite a little delegate interface, then write an implementation that we pass the thread pool/coroutine context/whatever, and the error reporter.
+
+```kotlin
+interface DemoDelegate {
+    fun onBackgroundThread(rustCall: () -> Unit)
+    fun <T> withErrorReporter(rustCall: () -> T): T
+    fun thenNotifyListeners(obj: DemoLib, rustCall: () -> T)
+}
+
+class MyDemoDelegate(
+    val backgroundScope: CoroutineContext,
+    val errorReporter: (e: Exception) -> Unit
+) : DemoDelegate {
+    fun onBackgroundThread(rustCall: () -> Unit) {
+        backgroundScope.launch {
+            rustCall()
+        }
+    }
```

Comment:
I think it would be useful to work an example of asyncifying a method call here - not just having it run on a background thread, but having it asynchronously return a value to the caller via whatever mechanism is appropriate for the foreign language in question. I suspect that handling the return types in a generic way may be challenging, but it'll definitely be interesting.

(I am assuming that we'd like to use this facility to replace [this kind of asyncification boilerplate](https://github.com/mozilla-mobile/android-components/blob/71e124a5dbafffb800168ea615fdbc54d7398f50/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FirefoxAccount.kt#L27) as well).

Reply:
I'm leaning towards removing the example around launching on a background thread, and leaving that to #1054 .

While this delegate/decorator pattern fits very well what we're doing with off the main thread work in some projects, its focus is to clean up boiler plate, not to solve the async problem. As such, talking about threading in this issue is a distraction, which has derailed the discussion.